clients/go/events/idempotent_recorder.go

```diff
@@ -0,0 +1,99 @@
+package events
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	eventsErr "github.com/lyft/flyteidl/clients/go/events/errors"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/event"
+
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/lyft/flytestdlib/promutils"
+)
+
+type EmittedEventInfo struct {
+	lastUpdatedAt time.Time
+	phase         string
+}
+
+type idempotentWorkFlowEventRecorder struct {
+	internalRecorder    *eventRecorder
+	events              map[string]EmittedEventInfo
+	maxUpdateLagSeconds int64
+	mutex               *sync.Mutex
+}
+
+func (i *idempotentWorkFlowEventRecorder) idempotentRecord(ctx context.Context, id string, e proto.Message, phase string, terminal bool) error {
```

Comment:
this wouldnt work right, when will you remove the events, Why not use LRU map that we have?

Reply:
we remove it here: https://github.com/lyft/flyteidl/pull/58/files#diff-88bad519713daa3e85d2e82c787119ccR47

basically on an terminal event. I actually had an LRU map for further keep the size in check for the case when really large number of WFs are running :) I removed it as it seemed like an unnecessary complication for our scenario (with explicit terminal state)